service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java

```diff
@@ -127,74 +147,43 @@ public synchronized MethodResult invoke(String deviceId, String moduleId, String
      * @param methodName is the name of the method that shall be invoked on the device.
      * @param responseTimeoutInSeconds is the maximum waiting time for a response from the device in seconds.
      * @param connectTimeoutInSeconds is the maximum waiting time for a response from the connection in seconds.
-     * @param payload is the the method parameter
-     * @return the status and payload resulted from the method invoke
-     * @throws IotHubException This exception is thrown if the response verification failed
-     * @throws IOException This exception is thrown if the IO operation failed
+     * @param payload is the the method parameter.
+     * @return the status and payload resulted from the method invoke.
+     * @throws IotHubException This exception is thrown if the response verification failed.
+     * @throws IOException This exception is thrown if the IO operation failed.
      */
     private synchronized MethodResult invokeMethod(URL url, String methodName, Long responseTimeoutInSeconds, Long connectTimeoutInSeconds, Object payload) throws IotHubException, IOException
     {
-        /* Codes_SRS_DEVICEMETHOD_21_006: [The invoke shall throw IllegalArgumentException if the provided responseTimeoutInSeconds is negative.] */
-        /* Codes_SRS_DEVICEMETHOD_21_007: [The invoke shall throw IllegalArgumentException if the provided connectTimeoutInSeconds is negative.] */
-        /* Codes_SRS_DEVICEMETHOD_21_014: [The invoke shall bypass the Exception if one of the functions called by invoke failed.] */
         MethodParser methodParser = new MethodParser(methodName, responseTimeoutInSeconds, connectTimeoutInSeconds, payload);
 
-        /* Codes_SRS_DEVICEMETHOD_21_011: [The invoke shall add a HTTP body with Json created by the `serializer.MethodParser`.] */
         String json = methodParser.toJson();
         if(json == null)
         {
-            /* Codes_SRS_DEVICEMETHOD_21_012: [If `MethodParser` return a null Json, the invoke shall throw IllegalArgumentException.] */
             throw new IllegalArgumentException("MethodParser return null Json");
         }
 
-        long  responseTimeout, connectTimeout;
+        Proxy proxy = options.getProxyOptions() != null ? options.getProxyOptions().getProxy() : null;
```

Comment:
If proxy is presesnt as a part of options, then in order to invoke methods and update twins, I'd have to set proxy to their specific "options" types, correct?
Wouldn't that be a bit of overhead? What do think about having a ClientOptions type instead (with proxy, and anything else that would be common across the client), and have these operations specific options inherit from that?

Reply:
Customers can re-use the same ```ProxyOptions``` instance, but you are right that they would have to create separate methods/twin/registry options classes to house that. It is a bit of overhead, sure, but I think this is the easiest design for customer's to understand.

I'm not sure inheritance here would make for the best user experience. If the constructor for a twin/method service client took a concrete class named ```ClientOptions```, then how am I supposed to know that there is also some ```TwinClientOptions``` class that inherits from it that has other options that I may want? I'd rather just have several concrete classes so that the API is quick to pick up and customer's won't miss important configurable options.

Reply:
Yes, I agree; inheritance will not be solving anything here from customer's perspective, but it might make code maintenance easier for us though.

Reply:
But what is more important, customer usability or code maintenance? I don't think that we are imposing too much on ourselves here